### PR TITLE
GH-111843: Tier 2 exponential backoff

### PR DESCRIPTION
**Pull request by markshannon**
Use exponential backoff to reduce the number of failed optimization attempts in tier 2.

[The stats](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20231105-3.13.0a1%2B-bff925c/bm-20231105-azure-x86_64-faster%252dcpython-tier_2_exponential_b-3.13.0a1%2B-bff925c-pystats.md) show that tier 2 optimization is now 75% effective. Previously it was 0.2% effective.

* Issue: gh-111843

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2335,19 +2335,29 @@ dummy_func(
...
                 ERROR_IF(optimized < 0, error);
                 if (optimized) {
                     // Rewind and enter the executor:
                     assert(this_instr->op.code == ENTER_EXECUTOR);
                     next_instr = this_instr;
+                    this_instr[1].cache &= ((1 << OPTIMIZER_BITS_IN_COUNTER) - 1);
+                }
+                else {
+                    int backoff = counter & ((1 << OPTIMIZER_BITS_IN_COUNTER) - 1);
+                    if (backoff < OPTIMIZER_BITS_IN_COUNTER + MINIMUM_TIER2_BACKOFF) {
```
Comment (iritkatriel) [inline Python/bytecodes.c]: why are you comparing backoff with OPTIMIZER_BITS_IN_COUNTER?

File: Python/pylifecycle.c [hunk 2]
```diff
@@ -1240,6 +1240,7 @@ init_interp_main(PyThreadState *tstate)
         if (_Py_get_xoption(&config->xoptions, L"uops") != NULL) {
             enabled = 1;
         }
+        enabled = 1;
```
Comment (corona10) [inline Python/pylifecycle.c]: I think that this is a hard-coded value for your experimentation, It should be removed right?
```suggestion
```

Comment (markshannon) [reply Python/pylifecycle.c]: Yes, thanks.

File: Python/optimizer.c [hunk 3]
```diff
@@ -107,6 +107,7 @@ error_optimize(
     _PyExecutorObject **exec,
     int Py_UNUSED(stack_entries))
 {
+    assert(0);
```
Comment (iritkatriel) [inline Python/optimizer.c:110]: ?

Comment (markshannon) [reply Python/optimizer.c:110]: So, it aborts on debug builds. Release builds will raise a `SystemError`

**Review by iritkatriel — APPROVED** [review]

**Pull request merged**